dali/operators/imgcodec/image_decoder.h

```diff
@@ -0,0 +1,761 @@
+    auto get_task = [&](int block_idx, int nblocks) {
+      return [&, block_idx, nblocks](int tid) {
+        int i_start = nsamples * block_idx / nblocks;
+        int i_end = nsamples * (block_idx + 1) / nblocks;
+        for (int i = i_start; i < i_end; i++) {
+          auto *st = state_[i].get();
+          assert(st != nullptr);
+          const auto &input_sample = input[i];
+
+          auto src_info = input.GetMeta(i).GetSourceInfo();
+          if (use_cache && cache_->IsInCache(src_info)) {
+            auto cached_shape = cache_->CacheImageShape(src_info);
+            auto roi = GetRoi(spec_, ws, i, cached_shape);
+            if (!roi.use_roi()) {
+              shapes.set_tensor_shape(i, cached_shape);
+              continue;
+            }
+          }
+          ParseSample(st->parsed_sample,
+                      span<const uint8_t>{static_cast<const uint8_t *>(input_sample.raw_data()),
+                                          volume(input_sample.shape())});
+          ROI &roi = rois_[i] = GetRoi(spec_, ws, i, st->parsed_sample.dali_img_info.shape);
+          OutputShape(shapes.tensor_shape_span(i), st->parsed_sample.dali_img_info, format_,
+                      is_planar_layout, use_orientation_, roi);
+        }
+      };
+    };
+
+    int nblocks = tp_->NumThreads() + 1;
+    if (nsamples > nblocks * 4) {
+      int block_idx = 0;
+      for (; block_idx < tp_->NumThreads(); block_idx++) {
+        tp_->AddWork(get_task(block_idx, nblocks), -block_idx);
+      }
+      tp_->RunAll(false);                // start work but not wait
+      get_task(block_idx, nblocks)(-1);  // run last block
+      tp_->WaitForWork();                // wait for the other threads
+    } else {                             // not worth parallelizing
+      get_task(0, 1)(-1);                // run all in current thread
+    }
+
+    output_descs[0] = {std::move(shapes), dtype_};
+    return true;
+  }
+
+  template <typename OutBackend>
+  void PrepareOutput(SampleState &st, SampleView<OutBackend> out, const ROI &roi,
+                     const Workspace &ws) {
+    st.image_info = st.parsed_sample.nvimgcodec_img_info;  // copy we modify
+    st.orig_layout = st.req_layout = "HWC";
+    st.req_img_type = format_;
+    auto info = st.parsed_sample.dali_img_info;
+    TensorShape<> decode_out_sh = info.shape;
+    int64_t &nchannels = decode_out_sh[2];
+
+    // Decode to format
+    if (format_ == DALI_ANY_DATA) {
+      st.image_info.sample_format = NVIMGCODEC_SAMPLEFORMAT_I_UNCHANGED;
+      st.orig_img_type = DALI_ANY_DATA;
+    } else if (nchannels == 1 || format_ == DALI_GRAY) {
+      st.image_info.sample_format = NVIMGCODEC_SAMPLEFORMAT_P_Y;
+      st.orig_img_type = DALI_GRAY;
+      nchannels = 1;
+    } else {
+      st.image_info.sample_format = NVIMGCODEC_SAMPLEFORMAT_I_RGB;
+      st.orig_img_type = DALI_RGB;
+      nchannels = 3;
+    }
+
+    st.image_info.color_spec = NVIMGCODEC_COLORSPEC_SRGB;
```

Comment:
What does it mean for non-rgb color spaces? In any case, I'd move this line before the if-else ladder, so we can possibly overwrite it in some cases.